### DISCUSSION TITLE
KTL Setup - How to disable features

### DISCUSSION THREAD

**Discussion by cortexrd** [cortexrd/Knack-Toolkit-Library#51] [Show and tell]
### Disabling KTL Features

Let's say you've just installed the KTL with the 5 lines of the Loader.  You've experimented a bit with all the basic features, but now, you want to change the default behavior and adapt the KTL to your needs.  Typically, you realize that some features are not desirable and you want to turn them off.

Here's how to do it.

You need to copy the content of this file to your Builder's Javascript pane, just below the Loader:
https://github.com/cortexrd/Knack-Toolkit-Library/blob/master/KTL_Customization.js

The KTL_Customization.js file is actually a very close copy of its big sister:
https://github.com/cortexrd/Knack-Toolkit-Library/blob/master/KTL_Defaults.js

It can be seen as a small "template" copy has just a few flags to get you started.

Note: all flags can be removed without problem if you want to keep your code as small as possible.  In that case, they will be initialized with the defaults as set in the KTL_Defaults.js file.

This code below has been stripped-of most of its code, and would still be perfectly valid and functional.

```
//Your app's version.
window.APP_VERSION = '1.0.0';

var ktl;
window.ktlReady = function (appInfo = {}) {
    ktl = window.ktl;

    //KTL Setup - BEGIN
    ktl.scenes.setCfg({
        versionDisplayName: 'MyApp', //This will show a short version name of your app.  Leave empty to remove it.
    })

    ktl.core.setCfg({
        developerNames: ['John Doe'], //Add your name here to get super powers!

        enabled: {
            versionInfo: {
                viShowAppInfo: true,
                viShowKtlInfo: true,
                viShowToRoles: ['Developer'], //Array of Account Roles strings who can see the version info bar. Leave empty for all.
                viPosX: 'right', //right, center, left
                viPosY: 'top', //top, bottom
                viPosXMobile: 'center', //right, center, left
                viPosYMobile: 'bottom', //top, bottom
                viOpacity: 10, //0 to 100
                viOpacityHover: 100, //0 to 100
            },
        },
    })

    //KTL Setup - END

    //KTL callbacks to your App - BEGIN

    //KTL callbacks to your App - END

    /////////////////////////////////////////////////
    //Your App-specific code goes here...

}

```

If you want to disable a feature **the flag must exists** and set its value to **false**.

### Examples:

1. I like everything, but I want to disable only the auto-focus
2. Locate the flag for that feature and set its value to **false**
3. Delete all the rest that you don't need, leaving the features enabled
4. The minimal setup code now looks like this:

```
//Your app's version.
window.APP_VERSION = '1.0.0';

var ktl;
window.ktlReady = function (appInfo = {}) {
    ktl = window.ktl;

    ktl.core.setCfg({
        enabled: {
            autoFocus: false,
        },
    })
}

```

**Comment by jbk68** [comment]
Hey Norm, I like having the ability to show my version info on some of my pages but would like the ability to select what pages this info is not shown on?
For example, show it on all view with the exception of "view_25, view_26"... JON

**Reply by cortexrd** [reply]
Hi Jon,

1. The version info bar is not shown per view, but per page.
2. There's already an option to show it only for specific roles - typically Developer - and not for all others.  Usually that solves the problem for most cases.
3. If what you need is to exclude on a per-scene basis, then some simple code can do it.  See below.

Copy the pieces of code at the right place in your own code - those lines between the `BEGIN` and `END` markers.
Replace `scene_1` and `scene_2 `by the real scene IDs.  Add or remove as needed.
The exlamation mark means "scenes must NOT be included".  If you remove it, it will apply the _normal_ logic: only "add the info bar IF the scenes are included" in the array (list).

```
window.ktlReady = function (appInfo = {}) {
    var ktl = new Ktl($, appInfo);


    //KTL Setup - BEGIN

    ktl.scenes.setCfg({
        onSceneRender: onSceneRender,
    })

    //KTL Setup - END




    //KTL callbacks to your App - BEGIN

    function onSceneRender(event, scene) {
        if (!['scene_1', 'scene_2'].includes(scene.key))
            ktl.scenes.addVersionInfo(appInfo);
    }

    //KTL callbacks to your App - END
}
```

Enjoy,
Norm

**Comment by jbk68** [comment]
Thank you... JON
